### PR TITLE
Media field

### PR DESCRIPTION
**Pull request by JeminiBo**

### CODE REVIEW

File: configs/field-types/Text/Text.jsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  const classes = useStyles();
+
+  return (
+    <div className={classes.wrapper}>
+      {label && <InputLabel>{}</InputLabel>}
+      <TextField
+        {...rest}
+        ref={forwardedRef}
+        value={!value && value !== 0 ? '' : value}
```
Comment (sergesemashko) [inline configs/field-types/Text/Text.jsx:35]: Didn’t quite get what case is handled here: if not value and value is not equal 0 then assign empty string. could you elaborate?

File: server/index.js [hunk 2]
```diff
@@ -37,8 +69,41 @@ app.prepare().then(() => {
...
 
+  server.delete('/_api/content/remove-media', (req, res) => {
+    return content.removeMedia(req.query.name, result => {
+      res.send(result);
+    });
+  });
+
+  server.post('/_api/content/save-media', (req, res) => {
+    const promise = new Promise((resolve, reject) => {
+      upload.array('file', 12)(req, res, err => {
```
Comment (sergesemashko) [inline server/index.js]: what is `12` exactly? can you define a constant for that?

File: server/index.js [hunk 3]
```diff
@@ -37,8 +69,41 @@ app.prepare().then(() => {
     });
   });
 
+  server.delete('/_api/content/remove-media', (req, res) => {
+    return content.removeMedia(req.query.name, result => {
+      res.send(result);
+    });
+  });
+
+  server.post('/_api/content/save-media', (req, res) => {
```
Comment (sergesemashko) [inline server/index.js]: `save` is redundant. post in REST is already means creation

File: server/index.js [hunk 4]
```diff
@@ -37,8 +69,41 @@ app.prepare().then(() => {
     });
   });
 
+  server.delete('/_api/content/remove-media', (req, res) => {
```
Comment (sergesemashko) [inline server/index.js]: remove is redundant, `delete` method already means removal

File: server/index.js [hunk 5]
```diff
@@ -20,11 +32,31 @@ app.prepare().then(() => {
       res.send(JSON.stringify(files));
     });
   });
+
+  server.get('/_api/content/get-media', (req, res) => {
```
Comment (sergesemashko) [inline server/index.js]: let’s make it `'/_api/content/media/:mediaUuid’`

File: server/index.js [hunk 6]
```diff
@@ -20,11 +32,31 @@ app.prepare().then(() => {
...
+        res.send(result);
+      });
+    } else {
+      return content.getMedia([req.query.file], result => {
+        res.send(result);
+      });
+    }
+  });
+
+  server.get('/_api/content/media-list', (req, res) => {
```
Comment (sergesemashko) [inline server/index.js]: let’s leave as just `/_api/content/media`

File: server/content.js [hunk 7]
```diff
@@ -27,6 +35,71 @@ const listTypes = cb => {
...
+        mediaPath = path.resolve(
+          process.env.MEDIA_FOLDER,
+          mediaData.currentDate,
+          mediaData.filename,
+        );
+
+        fs.stat(mediaPath, (err, stat) => {
+          if (err === null) {
+            mediaList.push(
+              `/${process.env.MEDIA_FOLDER}/${mediaData.currentDate}/${mediaData.filename}`,
```
Comment (sergesemashko) [inline server/content.js]: 👍

File: server/content.js [hunk 8]
```diff
@@ -27,6 +35,71 @@ const listTypes = cb => {
...
+      });
+    });
+
+    Promise.all(promises).then(() => cb(filesArray));
+  });
+};
+
+const getMedia = (uuidArray, cb) => {
+  const mediaList = [];
+  console.log(uuidArray);
```
Comment (sergesemashko) [inline server/content.js]: let’s remove console.log

File: package.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
   "description": "Flat file NodeJS based CMS",
   "main": "index.js",
   "scripts": {
-    "dev": "cross-env NODE_ENV=development nodemon server.js --watch server.js",
+    "dev": "cross-env NODE_ENV=development nodemon -r dotenv/config server.js --watch server.js",
```
Comment (sergesemashko) [inline package.json:7]: 👍

File: core/ui/mediaLibrary/index.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  };
+
+  return (
+    <Modal
+      isOpen={modalIsOpen}
+      onAfterOpen={() => mediaList().then(data => setMediaArray(data))}
+      onRequestClose={() => setModalIsOpen(false)}
+      contentLabel="Example Modal"
+      style={{
+        content: { top: '15%', left: '15%', right: '15%', bottom: '15%', overflow: 'visible' },
```
Comment (sergesemashko) [inline core/ui/mediaLibrary/index.js]: can we make it a CSS class?

File: core/client/api/content.js [hunk 11]
```diff
@@ -7,6 +8,33 @@ const load = async (contentType, slug) => {
   return response.data;
 };
 
+const loadMedia = async names => {
+  const response = await axios.get(`${HOST}/_api/content/get-media`, {
```
Comment (sergesemashko) [inline core/client/api/content.js]: let’s update the endpoints as per comments in the server/index.js file

File: core/ui/mediaLibrary/index.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    files.forEach(file => {
+      data.append('file', file);
+    });
+    saveMedia(data);
+    mediaList().then(data => setMediaArray(data));
+  };
+
+  const Library = () => {
+    const cards = mediaArray.map((media, index) => (
+      <div
```
Comment (sergesemashko) [inline core/ui/mediaLibrary/index.js:112]: looks like this snippet deserves a separate component

File: server/index.js [hunk 13]
```diff
@@ -5,13 +5,25 @@ const port = parseInt(process.env.PORT, 10) || 3000;
 const dev = process.env.NODE_ENV !== 'production';
 const app = next({ dev });
 const handle = app.getRequestHandler();
+const multer = require('multer');
+const uuidv4 = require('uuid/v4');
 
 const content = require('./content');
 const contentType = require('./content-type');
+const currentDate = `${new Date().getMonth() +
```
Comment (sergesemashko) [inline server/index.js]: this const should cannot be in the global scope as it’s going to be dynamic per request. once it’s defined in global scope it’ll only refresh on full server restart

**Review by sergesemashko — CHANGES_REQUESTED** [review]
Looks really good.
A few comments:
- I’m getting an error on some content items
Screen Shot 2019-07-21 at 9 40 32 PM
- there is some image flickering in the library when hovering over the images
- Let’s avoid stretching the image to fill the container. Let’s try out [object-fit:cover](https://css-tricks.com/almanac/properties/o/object-fit/) to let the image preserve it’s proportions in the square 
Screen Shot 2019-07-21 at 9 42 25 PM

- let’s align image to the left in image library rather than centering those

File: server/index.js (hunk 2 above)
Comment (JeminiBo) [reply server/index.js]: Max files count

File: configs/field-types/Text/Text.jsx (hunk 1 above)
Comment (JeminiBo) [reply configs/field-types/Text/Text.jsx:35]: Need clarify it with @k-stasevich

File: server/index.js [hunk 14]
```diff
@@ -5,13 +5,29 @@ const port = parseInt(process.env.PORT, 10) || 3000;
...
 const app = next({ dev });
 const handle = app.getRequestHandler();
+const multer = require('multer');
+const uuidv4 = require('uuid/v4');
 
 const content = require('./content');
 const contentType = require('./content-type');
 
+const createStorage = () => {
+  const currentDate = `${new Date().getMonth() +
```
Comment (sergesemashko) [inline server/index.js:15]: again, `createStorage()` is called only once on bootstrap thus currentDate is generated only once too and not going to change over the time. You probably want to prepend the date to the filename:
```js
filename(req, file, cb) {
   const currentDate = `${new Date().getMonth() + 1}.${new Date().getDate()}.${new Date().getFullYear()}`;
   cb(null, path.join(currentDate, file.originalname));
  },

```

File: core/client/api/content.js (hunk 11 above)
Comment (sergesemashko) [reply core/client/api/content.js]: let's have just `..._api/content/media`

File: server/index.js (hunk 6 above)
Comment (sergesemashko) [reply server/index.js]: don't forget to update this

File: server/index.js [hunk 15]
```diff
@@ -20,11 +36,31 @@ app.prepare().then(() => {
       res.send(JSON.stringify(files));
     });
   });
+
+  server.get('/_api/content/media', (req, res) => {
```
Comment (sergesemashko) [inline server/index.js]: `/_api/content/media/{uuid}` GET|PUT|PATCH|DELETE
`/_api/content/media?filters[uuid]=['']` GET|POST

**Comment by sergesemashko** [comment]
looks good, really great job :+1:

**Pull request merged**